src/bp/core/services/dialog/prompt-manager.ts

```diff
@@ -0,0 +1,238 @@
+      await this.actionStrategy.invokeSendMessage(buildMessage(node.question), '@builtin_text', event)
+    }
+  }
+
+  private async _askConfirmation(event: IO.IncomingEvent, value: any, node: PromptNode) {
+    debugPrompt('low confidence, asking validation for %o', { value: value, output: node.output })
+
+    const confirmNode = getConfirmPromptNode(node, value)
+    await this._sendCustomPrompt(event, new PromptConfirm(confirmNode), confirmNode)
+  }
+
+  private async _sendCustomPrompt(incomingEvent: IO.IncomingEvent, prompt: Prompt, node: PromptNode) {
+    debugPrompt('sending custom prompt to user')
+
+    const promptEvent = Event({
+      ..._.pick(incomingEvent, ['botId', 'channel', 'target', 'threadId']),
+      direction: 'outgoing',
+      type: 'prompt',
+      payload: node,
+      incomingEventId: incomingEvent.id
+    })
+
+    const bp = await createForBotpress()
+    await prompt.customPrompt?.(promptEvent, incomingEvent, bp)
+
+    await this.eventEngine.sendEvent(promptEvent)
+  }
+
+  private async _continueOriginalEvent(event: IO.IncomingEvent) {
+    const { originalEvent } = event.state.session.prompt!
+    const promptEvent = Event(originalEvent as IO.IncomingEvent) as IO.IncomingEvent
+
+    promptEvent.restored = true
+    promptEvent.state = _.omit(event.state, ['session.prompt']) as IO.EventState
+
+    await this.eventEngine.sendEvent(promptEvent)
+  }
+
+  /** Returns the last 6 events if it's the first time the prompt is executed */
+  private async _getLastEvents(event: IO.IncomingEvent): Promise<IO.IncomingEvent[]> {
+    if (event.state.session.prompt) {
+      return [event]
+    }
+
+    const lastEvents: IO.StoredEvent[] = await this.eventRepository.findEvents(
+      { direction: 'incoming', target: event.target },
+      { count: 5, sortOrder: [{ column: 'createdOn', desc: true }] }
+    )
+
+    return lastEvents.map(x => x.event as IO.IncomingEvent)
+  }
+
+  private _getPrompt(node: PromptNode, event: IO.IncomingEvent) {
+    const status = event.state.session?.prompt?.status
+
+    return this.loadPrompt(status?.confirming ? getConfirmPromptNode(node, status?.value) : node)
+  }
+
+  private loadPrompt(promptNode: PromptNode): { prompt: Prompt } & PromptConfig {
+    const promptClass = prompts.find(x => x.config.type === promptNode.type)
+    if (!promptClass) {
+      throw new Error(`Unknown prompt type ${promptNode.type}`)
+    }
+
+    const prompt = new promptClass(promptNode.params as any)
+    return { prompt, ...promptClass.config }
+  }
+
+  private async evaluateEventVariables(
+    events: IO.IncomingEvent[],
+    prompt: Prompt
+  ): Promise<{ confidence: number; extracted: any }[]> {
+    return Promise.mapSeries(events, async (event, idx) => {
+      const { value, confidence } = prompt.extraction(event) || {}
+      const { valid } = (await prompt.validate(value)) || {}
+      const finalConfidence = +!!valid * (confidence ?? 0) * (1 - idx * OLD_MESSAGE_CONFIDENCE_DECREASE)
```

Comment:
I think OLD_MESSAGE_CONFIDENCE_DECREASE intuition is good but I think this should be applied to the first previous message.

from what I've seen so far it's not

Reply:
Yup I agree